icu4c/source/i18n/basictz.cpp

```diff
@@ -293,71 +293,77 @@ BasicTimeZone::getTimeZoneRulesAfter(UDate start, InitialTimeZoneRule*& initial,
     }
 
     const InitialTimeZoneRule *orgini;
-    const TimeZoneRule **orgtrs = NULL;
     TimeZoneTransition tzt;
-    UBool avail;
-    UVector *orgRules = NULL;
+    bool avail;
     int32_t ruleCount;
-    TimeZoneRule *r = NULL;
-    UBool *done = NULL;
-    InitialTimeZoneRule *res_initial = NULL;
-    UVector *filteredRules = NULL;
+    TimeZoneRule *r = nullptr;
     UnicodeString name;
     int32_t i;
     UDate time, t;
-    UDate *newTimes = NULL;
     UDate firstStart;
-    UBool bFinalStd = FALSE, bFinalDst = FALSE;
+    UBool bFinalStd = false, bFinalDst = false;
+
+    initial = nullptr;
+    transitionRules = nullptr;
 
     // Original transition rules
     ruleCount = countTransitionRules(status);
     if (U_FAILURE(status)) {
         return;
     }
-    orgRules = new UVector(ruleCount, status);
+    LocalPointer<UVector> orgRules(
+        new UVector(uprv_deleteUObject, nullptr, ruleCount, status), status);
     if (U_FAILURE(status)) {
         return;
     }
-    orgtrs = (const TimeZoneRule**)uprv_malloc(sizeof(TimeZoneRule*)*ruleCount);
-    if (orgtrs == NULL) {
+    LocalMemory<const TimeZoneRule *>orgtrs(
+        static_cast<const TimeZoneRule **>(uprv_malloc(sizeof(TimeZoneRule*)*ruleCount)));
+    if (orgtrs.isNull()) {
         status = U_MEMORY_ALLOCATION_ERROR;
-        goto error;
+        return;
     }
-    getTimeZoneRules(orgini, orgtrs, ruleCount, status);
+    getTimeZoneRules(orgini, &orgtrs[0], ruleCount, status);
     if (U_FAILURE(status)) {
-        goto error;
+        return;
     }
     for (i = 0; i < ruleCount; i++) {
-        orgRules->addElementX(orgtrs[i]->clone(), status);
+        LocalPointer<TimeZoneRule>lpRule(orgtrs[i]->clone(), status);
+        orgRules->adoptElement(lpRule.orphan(), status);
         if (U_FAILURE(status)) {
-            goto error;
+            return;
         }
     }
-    uprv_free(orgtrs);
-    orgtrs = NULL;
 
     avail = getPreviousTransition(start, TRUE, tzt);
     if (!avail) {
         // No need to filter out rules only applicable to time before the start
         initial = orgini->clone();
-        transitionRules = orgRules;
+        if (initial == nullptr) {
+            status = U_MEMORY_ALLOCATION_ERROR;
+            return;
+        }
+        transitionRules = orgRules.orphan();
         return;
     }
 
-    done = (UBool*)uprv_malloc(sizeof(UBool)*ruleCount);
-    if (done == NULL) {
+    LocalMemory<bool>done(static_cast<bool *>(uprv_malloc(sizeof(UBool)*ruleCount)));
```

Comment:
I think we might want to use `sizeof(bool)`, since `UBool` and `bool` might have different sizes. (Probably very rare... but strictly speaking, `sizeof(bool)` is I think implementation defined, and `UBool` is `int8_t`).
```suggestion
    LocalMemory<bool>done(static_cast<bool *>(uprv_malloc(sizeof(bool)*ruleCount)));
```

Reply:
Good catch!
Done.